lib/ansible/modules/network/iosxr/iosxr_vrf.py

```diff
@@ -0,0 +1,319 @@
+            interfaces.append(interface)
+
+    return interfaces
+
+
+def map_config_to_obj(module):
+    output = run_commands(module, ['show vrf all'])
+    lines = output[0].strip().splitlines()[1:]
+
+    if not lines:
+        return list()
+
+    objs = list()
+
+    for l in lines:
+        obj = {}
+        splitted_line = re.split(r'\s{2,}', l.strip())
+        obj['name'] = splitted_line[0]
+        obj['interfaces'] = parse_interface(module, obj['name'])
+        objs.append(obj)
+
+    return objs
+
+
+def check_declarative_intent_params(want, module):
+    if module.params['interfaces']:
+        time.sleep(module.params['delay'])
+
+        for w in want:
+            command = 'show vrf {} detail'.format(w['name'])
+            rc, out, err = exec_command(module, command)
+            if rc != 0:
+                module.fail_json(msg=to_text(err, errors='surrogate_then_replace'), command=command, rc=rc)
+
+            have_interfaces = list()
+            match = re.search(r'Interfaces:(\n \S+)*', out, re.M)
+            if match:
+                interfaces = match.group(0).strip().split('\n  ')
+                if 'Interfaces' in interfaces[0]:
+                    have_interfaces = interfaces[1:]
+
+            for i in w['interfaces']:
+                if have_interfaces and i not in have_interfaces:
+                    module.fail_json(msg="Interface %s not configured on vrf %s" % (i, w['name']))
+
+
+def main():
+    """ main entry point for module execution
+    """
+    element_spec = dict(
+        name=dict(),
+        interfaces=dict(type='list'),
+        delay=dict(default=10, type='int'),
+        state=dict(default='present', choices=['present', 'absent'])
+    )
+
+    aggregate_spec = deepcopy(element_spec)
+    aggregate_spec['name'] = dict(required=True)
+
+    # remove default in aggregate spec, to handle common arguments
+    remove_default_spec(aggregate_spec)
+
+    argument_spec = dict(
+        aggregate=dict(type='list', elements='dict', options=aggregate_spec),
+        purge=dict(default=False, type='bool')
+    )
+
+    argument_spec.update(element_spec)
+    argument_spec.update(iosxr_argument_spec)
+
+    required_one_of = [['name', 'aggregate']]
+    mutually_exclusive = [['name', 'aggregate']]
+    module = AnsibleModule(argument_spec=argument_spec,
+                           required_one_of=required_one_of,
+                           mutually_exclusive=mutually_exclusive,
+                           supports_check_mode=True)
+
+    warnings = list()
+    check_args(module, warnings)
+
+    result = {'changed': False}
+    result['warnings'] = warnings
+
+    want = map_params_to_obj(module)
+    have = map_config_to_obj(module)
+
+    commands = map_obj_to_commands((want, have), module)
+    result['commands'] = commands
+
+    if commands:
+        if not module.check_mode:
+            load_config(module, commands, result['warnings'], commit=True)
```

Comment:
I think we should use `cliconf` connection plugin for new modules. I am working on updates to module_utils/iosxr.py. This can be refactored a bit for cliconf after that?